### PR TITLE
Add CuDNN RNN support

### PR DESCRIPTION
**Pull request by reuben**

### CODE REVIEW

**Comment by reuben** [comment]
Just checking if tests explode. Needs more profiling as well as updating our TensorFlow cwise rule.

**Comment by reuben** [comment]
This PR has one caveat: CuDNN and non-CuDNN training graphs aren't compatible. This means you can't start a training run with `--use_cudnn_rnn` and then continue it on the CPU, or on a GPU but without CuDNN RNN. I'm trying to come up with a clean way to fix it but in the mean time this PR can be reviewed independently, as it doesn't change any defaults.

**Comment by reuben** [comment]
Upon closer inspection, I don't think fixing the caveat in a bullet proof way is possible with the CudnnCompatible* machinery in TF. It seems geared exclusively for exporting models, basically a last step you do when you're done training. If you try to load a CudnnLSTM checkpoint in a CudnnCompatibleLSTMCell graph you get errors about the Adam tensors not being there.

A possible workaround would be to re-initialize the Adam moment tensors from scratch if they're missing from the checkpoint, but it could interfere significantly with transfer learning and fine tuning.

In summary: using CudnnRNN greatly speeds up training on GPUs, which would be great for our experiments and training runs on large datasets. But if we release CudnnRNN checkpoints, it reduces the audience that can make use of it, as it requires a CUDA GPU. @kdavis-mozilla @lissyx @tilmankamp thoughts?

**Comment by dabinat** [comment]
Does this produce models that will work for inference on the 0.5.1 native client? Or does the client need to be updated to work with models created via this method?

**Comment by lissyx** [comment]
> But if we release CudnnRNN checkpoints, it reduces the audience that can make use of it, as it requires a CUDA GPU

Ouch, that's going to be complicated. I often re-use checkpoints on my laptop to perform exports :/

**Comment by reuben** [comment]
> Does this produce models that will work for inference on the 0.5.1 native client? Or does the client need to be updated to work with models created via this method?

We broke compatibility with 0.5.1 with the fix for concurrent streams with a single ModelState. But this PR alone does not break compatibility with the clients.

> Ouch, that's going to be complicated. I often re-use checkpoints on my laptop to perform exports :/

Exporting works fine, it's just continuing training from the checkpoint that's problematic.

**Comment by bernardohenz** [comment]
@reuben , I remember that a time ago you could continuing training from a pb file (instead of a checkpoint). Couldn't this solve this issue (but losing information about optimizer probably )?

Complementing the question of @lissyx for checking, could a non-GPU notebook export a pb file from a CudnnRNN checkpoint?

**Comment by reuben** [comment]
> @reuben , I remember that a time ago you could continuing training from a pb file (instead of a checkpoint). Couldn't this solve this issue (but losing information about optimizer probably )?

This would be equivalent to my proposed workaround, just initialize the missing optimizer tensors from scratch.

> Complementing the question of @lissyx for checking, could a non-GPU notebook export a pb file from a CudnnRNN checkpoint?

Yep.

**Comment by reuben** [comment]
Forgot Tilman is on PTO :)

**Comment by lissyx** [comment]
> > Complementing the question of @lissyx for checking, could a non-GPU notebook export a pb file from a CudnnRNN checkpoint?
> 
> Yep.

That still means we block re-training with anything else than CUDA.

How much is the speedup?

File: DeepSpeech.py [hunk 1]
```diff
@@ -79,32 +79,65 @@ def dense(name, x, units, dropout_rate=None, relu=True):
 
 
 def rnn_impl_lstmblockfusedcell(x, seq_length, previous_state, reuse):
-    # Forward direction cell:
-    fw_cell = tf.contrib.rnn.LSTMBlockFusedCell(Config.n_cell_dim, reuse=reuse)
+    with tf.variable_scope('cudnn_lstm/rnn/multi_rnn_cell/cell_0'):
```
Comment (lissyx) [inline DeepSpeech.py]: I'm not sure I get what this is supposed to do precisely. You place `LSTMBlockCell` below a `cudnn_lstm` ? Or you **replace** ?

File: DeepSpeech.py [hunk 2]
```diff
@@ -79,32 +79,65 @@ def dense(name, x, units, dropout_rate=None, relu=True):
...
+    return output, output_state
+
+# Hack: CudnnLSTM works similarly to Keras layers in that when you instantiate
+# the object it creates the variables, and then you just call it several times
+# to enable variable re-use. Because all of our code is structure in an old
+# school TensorFlow structure where you can just call tf.get_variable again with
+# reuse=True to reuse variables, we can't easily make use of the object oriented
+# way CudnnLSTM is implemented, so we save a singleton instance in the function,
+# emulating a static function variable.
+rnn_impl_cudnn_rnn.cell = None
```
Comment (lissyx) [inline DeepSpeech.py]: I get the explanation, maybe it is worth explaining where the solution from the hack comes from? Reference to some source code ?

File: DeepSpeech.py [hunk 3]
```diff
@@ -79,32 +79,65 @@ def dense(name, x, units, dropout_rate=None, relu=True):
...
-    # need to do different things here.
-    output, output_state = tf.nn.static_rnn(cell=fw_cell,
-                                            inputs=x,
-                                            initial_state=previous_state,
-                                            dtype=tf.float32,
-                                            sequence_length=seq_length)
-    output = tf.concat(output, 0)
+
+def rnn_impl_static_rnn(x, seq_length, previous_state, reuse):
+    with tf.variable_scope('cudnn_lstm/rnn/multi_rnn_cell'):
```
Comment (lissyx) [inline DeepSpeech.py]: Same question as earlier, is this placing under this level, or exposing / replacing ?

File: util/flags.py [hunk 4]
```diff
@@ -51,9 +51,11 @@ def create_flags():
...
     f.DEFINE_integer('export_batch_size', 1, 'number of elements per batch on the exported graph')
 
-    # Performance(UNSUPPORTED)
-    f.DEFINE_integer('inter_op_parallelism_threads', 0, 'number of inter-op parallelism threads - see tf.ConfigProto for more details')
-    f.DEFINE_integer('intra_op_parallelism_threads', 0, 'number of intra-op parallelism threads - see tf.ConfigProto for more details')
+    # Performance
+
+    f.DEFINE_integer('inter_op_parallelism_threads', 0, 'number of inter-op parallelism threads - see tf.ConfigProto for more details. USE OF THIS FLAG IS UNSUPPORTED')
+    f.DEFINE_integer('intra_op_parallelism_threads', 0, 'number of intra-op parallelism threads - see tf.ConfigProto for more details. USE OF THIS FLAG IS UNSUPPORTED')
+    f.DEFINE_boolean('use_cudnn_rnn', False, 'use CuDNN RNN backend for training on GPU')
```
Comment (lissyx) [inline util/flags.py]: Maybe worth documenting that this will produce a checkpoint that cannot be trained on anything else than CUDA.

File: DeepSpeech.py [hunk 5]
```diff
@@ -632,7 +670,7 @@ def create_inference_graph(batch_size=1, n_steps=16, tflite=False):
     }
 
     if not FLAGS.export_tflite:
-        inputs.update({'input_lengths': seq_length})
+        inputs['input_lengths'] = seq_length
```
Comment (lissyx) [inline DeepSpeech.py]: Why did we put `update` in the first place? I can't remember

File: DeepSpeech.py [hunk 6]
```diff
@@ -573,7 +611,7 @@ def create_inference_graph(batch_size=1, n_steps=16, tflite=False):
 
     if batch_size <= 0:
         # no state management since n_step is expected to be dynamic too (see below)
-        previous_state = previous_state_c = previous_state_h = None
+        previous_states = None
```
Comment (lissyx) [inline DeepSpeech.py]: I don't see where `previous_states` is being used

File: DeepSpeech.py [hunk 7]
```diff
@@ -659,19 +697,16 @@ def export():
...
-        # a static RNN with a normal cell, so we need to rewrite the names to
-        # match the training weights when restoring.
-        def fixup(name):
-            if name.startswith('rnn/lstm_cell/'):
-                return name.replace('rnn/lstm_cell/', 'lstm_fused_cell/')
-            return name
-
-        mapping = {fixup(v.op.name): v for v in tf.global_variables()}
+    # Create a saver using variables from the above newly created graph
+    # Training graph uses LSTMFusedCell, but the TFLite inference graph uses
```
Comment (lissyx) [inline DeepSpeech.py]: nit: training can use CUDNN cells as well now

**Comment by reuben** [comment]
> That still means we block re-training with anything else than CUDA.

Yes, it's unfortunate.

> How much is the speedup?

Usually around 2x faster time per epoch.

File: DeepSpeech.py (hunk 2 above)
Comment (reuben) [reply DeepSpeech.py]: I don't understand. This comment was supposed to explain the solution. What's not clear? It doesn't come from anywhere, it's just a way to only instantiate the layer once.

File: DeepSpeech.py (hunk 5 above)
Comment (reuben) [reply DeepSpeech.py]: I don't know either. Maybe we were adding more than one key at once?

File: DeepSpeech.py (hunk 6 above)
Comment (reuben) [reply DeepSpeech.py]: This is a typo from a previous modification that I removed. Nice catch.

File: DeepSpeech.py (hunk 7 above)
Comment (reuben) [reply DeepSpeech.py]: Hm... Thinking more about this, I'm not sure this mapping is even needed anymore. Let me test it better.

File: DeepSpeech.py (hunk 2 above)
Comment (lissyx) [reply DeepSpeech.py]: @reuben Maybe I just did not got the point of the comment then. The comment itself is clear, but I don't get the link between the comment and this code `.cell = None`.

Comment (lissyx) [reply DeepSpeech.py]: Ok, re-reading again, I get the point, it's playing along with `if not rnn_impl_cudnn_rnn.cell` above. I guess it's just the placement that distrubs me. Moving it before `def rnn_impl_cudnn_rnn` might make it more obvious? Or I just lack sleep.
